grug/malachite/src/actors/host/host.rs

```diff
@@ -0,0 +1,560 @@
+                .call(|reply| MempoolMsg::Take { reply }, None)
+                .await?
+                .success_or(anyhow::anyhow!("Failed to take txs from mempool"))?;
+
+            // Call prepare_proposal
+            info!("Calling prepare_proposal");
+            let txs = state.prepare_proposal(txs);
+
+            let pre_block = PreBlock::new(
+                height,
+                self.address,
+                round,
+                Timestamp::from_nanos(
+                    std::time::SystemTime::now()
+                        .duration_since(std::time::UNIX_EPOCH)
+                        .unwrap()
+                        .as_nanos(),
+                ),
+                txs,
+            );
+
+            info!("Calling finalize_block");
+            let block_hash = state.finalize_block(&pre_block)?;
+
+            pre_block.into_block(block_hash)
+        };
+
+        let proposal_data = block.as_proposal_data();
+
+        let stream_id = state.stream_id();
+
+        // Store undecided proposal
+        let proposed_value = ProposedValue {
+            height,
+            round,
+            valid_round: Round::Nil,
+            proposer: self.address,
+            value: <ctx!(Value)>::new(proposal_data.block.block_hash()),
+            validity: Validity::Valid,
+        };
+
+        UNDECIDED_PROPOSALS.save(
+            state,
+            (*height, round.as_i64(), proposal_data.block.block_hash()),
+            &proposed_value,
+        )?;
+
+        // Store undecided block
+        UNDECIDED_BLOCK.save(state, proposal_data.block.block_hash(), &block)?;
+
+        let value = LocallyProposedValue::new(
+            height,
+            round,
+            <ctx!(Value)>::new(proposal_data.block.block_hash()),
+        );
+
+        // Stream parts to consensus
+        self.stream_data(stream_id, proposal_data, true).await?;
+
+        // Return the proposed value
+        reply_to.send(value)?;
+
+        Ok(())
+    }
+
+    #[tracing::instrument("received_proposal_part", skip_all)]
+    async fn received_proposal_part(
+        &self,
+        state: &mut State,
+        part: StreamMessage<ctx!(ProposalPart)>,
+        reply_to: RpcReplyPort<ProposedValue<Context>>,
+    ) -> ActorResult<()> {
+        // Since we are using a single part proposal, we only need data
+        let Some(ProposalData { block, valid_round }) = part.content.into_data() else {
+            return Ok(());
+        };
+
+        info!(height = %block.height, round = %block.round, "All parts received");
+
+        // Run FinalizeBlock
+        let block_hash = state.finalize_block(&block)?;
+
+        let value = ProposedValue {
+            height: block.height,
+            round: block.round,
+            valid_round,
+            proposer: block.proposer,
+            // TODO: Is this correct? If block hashes are different, which one we should use?
+            value: <ctx!(Value)>::new(block.block_hash()),
```

Comment:
It is correct, should be one from the received block and not the one you compute locally (`block_hash`)